### PR TITLE
Fix doubled lm_head gradient in Unsloth_Offloaded_Log_Softmax backward

### PR DESCRIPTION
**Pull request by danielhanchen**
### What

`Unsloth_Offloaded_Log_Softmax` (the CPU-offloading / checkpointed log-softmax used for the large / long-completion GRPO logprob path via `efficient_log_softmax`) recomputes its output in backward, then did:

```python
torch.autograd.backward(output, grad_output)
return (hidden_states.grad, lm_head.grad if ctx.lm_head_requires_grad else None, ...)
```

`torch.autograd.backward` writes into the leaf `.grad` attributes. Returning `lm_head.grad` as this op's gradient contribution then gets accumulated **again** by the outer engine's `AccumulateGrad`, so:

- lm_head gradient is **2x** too large for a single use,
- **more** when `lm_head` is shared across chunks (the normal GRPO path -- the same output embedding feeds every `compute_logprobs_chunk` call), and
- any gradient already accumulated on `lm_head` from a previous micro-batch is **corrupted** (returned as `2*pre + 2*g` instead of `pre + g`).

### Fix

Use `torch.autograd.grad(output, inputs, grad_output)`, which returns each input's **local** gradient contribution without touching leaf `.grad`. `hidden_states.grad` was already correct (its reloaded tensor is a private leaf) and is unchanged.

### Reachability

Only on the offload path (`index.shape[1] > 1024` or `batch_size > 8`, i.e. long-completion / large-batch GRPO) **and** with a trainable `lm_head` / tied embeddings (full fine-tuning, or LoRA with `modules_to_save=["lm_head"]`). Standard LoRA freezes `lm_head`, so the old path returned `None` and was unaffected.

### Verification

Reproduced against the non-offloaded twin `chunked_hidden_states_selective_log_softmax` as oracle: before the fix, `|lm_head.grad| / |ref|` was exactly 2.0 (single use) and 3.0 (two shared uses), with pre-existing grad corrupted to `2*pre + 2*g`; after the fix all cases match the oracle exactly (ratio 1.0, max abs diff 0.0), and `hidden_states.grad` stays exact. Added a CPU regression test that pins the pattern (fixed matches plain-autograd oracle for single / shared / pre-existing-grad; the old pattern provably diverges) plus a source guard.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -1389,11 +1389,20 @@ def backward(ctx, grad_output):
...
+            # Use torch.autograd.grad (not torch.autograd.backward) so we return each
+            # input's LOCAL gradient contribution. torch.autograd.backward writes into the
+            # leaf .grad attributes; returning lm_head.grad would then be double-counted by
+            # the outer engine's AccumulateGrad (2x for a single use, more when lm_head is
+            # shared across chunks) and would corrupt any grad already accumulated on lm_head.
+            grad_inputs = torch.autograd.grad(
+                output,
+                (hidden_states, lm_head) if ctx.lm_head_requires_grad else (hidden_states,),
+                grad_output,
+            )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1448]: This Function keeps the default materialize_grads=True (it never calls ctx.set_materialize_grads(False)), so PyTorch materializes any unused-output gradient as a zero tensor before calling backward. grad_output is therefore never None here, and the single output (the per-token logprobs) always feeds the loss regardless. A None guard would be unreachable dead code, so leaving it out.

**Comment by danielhanchen** [comment]
Commit 5c9f3646 is a comment-only cleanup (verified AST-identical code, tests unchanged at 18 passing).

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 2]
```diff
@@ -1353,9 +1353,43 @@ def forward(ctx, hidden_states, lm_head, index, chunks,
...
+            saved_hidden_states = None
+            if detached_hidden_states.is_cuda:
+                try:
+                    free_bytes, _ = torch.cuda.mem_get_info(detached_hidden_states.device)
+                    tensor_bytes = detached_hidden_states.numel() * detached_hidden_states.element_size()
+                except Exception:
+                    free_bytes, tensor_bytes = 0, 1
+                if tensor_bytes * 4 <= free_bytes:
+                    # Enough headroom: keep the alias on GPU (mem_get_info failure defaults to offload).
+                    saved_hidden_states = detached_hidden_states
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Good catch. Fixed in 05d9c63e: the keep decision now budgets cumulatively, 4 * (tensor_bytes + retained_this_step) <= free_bytes, capping retention at a quarter of measured free memory with at least three quarters left for the backward recompute. The counter lives in grpo_accumulated_loss scope and is recreated on every call, so no state can leak across steps. Verified with a 16 chunk retention test (old head kept 11 of 16 chunks, fixed head kept 3, all gradients bitwise exact) and the keep-on-GPU perf win is unchanged within noise.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Revisited this with measurements and you were more right than my previous reply gave you credit for. The cumulative budget removed the compounding across chunks but not the underlying mis-estimate: reserving 75% of free memory for the backward recompute is not enough, because the recompute peak scales with vocab size and the in-kernel chunk count. On B200 (bf16, B=8, H=4096) the minimum free VRAM needed to complete a step went up against always offloading:

  chunks=4  T=8192  V=32000   trainable    7.6 GB -> 8.8 GB  (+15.7%)
  chunks=4  T=8192  V=128000  frozen      23.6 GB -> 26.6 GB (+12.6%)
  chunks=16 T=2048  V=128000  trainable    9.3 GB -> 10.5 GB (+12.7%)

Two configurations out of a 120-config sweep completed with always-offload and OOMed with retention, reproducibly. In one of them the recompute alone already needed about 98% of free memory, so any retention at all was fatal.

Retention was worth about 2.9% throughput on top of the async copy, which is not a good trade on a path that only runs when the caller is already memory bound. Removed in 78829393; hidden states are now always offloaded. The async pinned side-stream D2H stays, since it is worth about 2.8% at identical peak GPU allocation. After the change peak allocation matches always-offload exactly and the figures above are back to parity.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 3]
```diff
@@ -1371,29 +1398,31 @@ def forward(ctx, hidden_states, lm_head, index, chunks,
...
                     hidden_states, lm_head, index, *ctx.args
                 )
 
-            torch.autograd.backward(output, grad_output)
+            # autograd.grad, not backward: backward writes into leaf .grad, which the
+            # outer AccumulateGrad would then double-count.
+            grad_inputs = torch.autograd.grad(
+                output,
+                (hidden_states, lm_head) if ctx.lm_head_requires_grad else (hidden_states,),
+                grad_output,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1447]: Confirmed and fixed in bc08233d. Reproduced against the non-offloaded twin with a scaling hook g -> g * 0.5 registered on lm_head:

  non-offloaded twin   hook fired 1x, |lm_head.grad| = 8.637692
  this PR (before fix) hook fired 2x, |lm_head.grad| = 4.318846   (0.500x)
  pre-PR implementation hook fired 2x, |lm_head.grad| = 12.956539 (1.500x)

So it predates this PR rather than being introduced by it, but the whole point of the Function is to match the twin, so it belongs here. The recompute now runs against a private lm_head.detach().requires_grad_(True) leaf, after which the hook fires once and the scaled gradient matches the twin exactly.

Confirming the mechanism, since it is unintuitive: the PyTorch autograd notes state that a hook registered with Tensor.register_hook is executed whenever a gradient is computed for that tensor, and that this does not require the tensor's grad_fn to run, specifically calling out the case where the tensor is passed in autograd.grad's inputs. Node-level hooks behave differently and only fire if their node executes, so DDP and FSDP, which hook AccumulateGrad rather than the tensor, were never double-firing here.

With no hook registered the gradients are unchanged and still bitwise equal to the twin for both leaf and non-leaf heads. Added a regression test that counts hook firings and checks the scaled gradient for both head kinds; it fails against the previous implementation.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 4]
```diff
@@ -1346,16 +1346,57 @@ def to_device(tensor, device, non_blocking=True):
...
+                try:
+                    pinned_buffer = torch.empty_like(detached_hidden_states, device = "cpu", pin_memory = True)
+                    if pinned_buffer is not None:
+                        current_stream = device_module.current_stream(detached_hidden_states.device)
+                        copy_stream = device_module.Stream(device = detached_hidden_states.device)
+                        copy_stream.wait_stream(current_stream)
+                        with device_module.stream(copy_stream):
+                            pinned_buffer.copy_(detached_hidden_states, non_blocking = True)
+                        # Keeps the GPU storage alive until the side-stream copy finishes.
+                        detached_hidden_states.record_stream(copy_stream)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1386]: Confirmed and fixed in fb0e72ae. You were right that the clone stays resident across the forward compute, and measuring it isolated the effect: my earlier memory sweep only tracked whole-step peak, which the backward recompute dominates, so this was invisible there. Measuring the forward phase alone on an 8 x 8192 x 4096 bf16 chunk (0.500 GiB clone):

  pre-PR (sync pageable)   1.175 GiB
  this PR before the fix   1.551 GiB   (+0.376 GiB)
  this PR after the fix    1.051 GiB   (parity)

The pre-PR code did not have the problem because its clone was a temporary consumed by the blocking .to("cpu") and became reclaimable before the log-softmax started, whereas the async version kept a named local alive for the rest of forward.

The fix is to drop the reference once the copy is enqueued rather than to wait on the event before computing. record_stream still prevents the allocator from handing that block out until the D2H lands, so the copy stays safe while the block becomes reusable mid-compute. Waiting instead would also fix the peak but would serialize the copy and remove the reason this path exists: pinned is +2.84% against pre-PR over four shapes after the fix, versus +2.77% before it, with a pageable control at +0.16%. Gradient matrix against the non-offloaded twin is unchanged at 1440 cases, 0 failures.

Added a test pinning that the release happens before the compute and that nothing touches the clone in between.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
